src/machine/machine_avr.go

```diff
@@ -271,3 +272,128 @@ func (i2c I2C) ReadByte() byte {
 
 	return byte(*avr.TWDR)
 }
+
+// UART
+
+type UARTConfig struct {
+	Baudrate uint32
+}
+
+type UART struct {
+}
+
+var (
+	// UART0 is the hardware serial port on the AVR.
+	UART0 = &UART{}
+)
+
+// Configure the UART on the AVR. Defaults to 9600 baud on Arduino.
+func (uart UART) Configure(config UARTConfig) {
+	if config.Baudrate == 0 {
+		config.Baudrate = 9600
+	}
+
+	// Set baud rate based on prescale formula from
+	// https://www.microchip.com/webdoc/AVRLibcReferenceManual/FAQ_1faq_wrong_baud_rate.html
+	// ((F_CPU + UART_BAUD_RATE * 8L) / (UART_BAUD_RATE * 16L) - 1)
+	ps := ((CPU_FREQUENCY+config.Baudrate*8)/(config.Baudrate*16) - 1)
+	*avr.UBRR0H = avr.RegValue(ps >> 8)
+	*avr.UBRR0L = avr.RegValue(ps & 0xff)
+
+	// enable RX interrupt
+	*avr.UCSR0B |= avr.UCSR0B_RXCIE0
+}
+
+// Close the UART on the AVR.
+func (uart UART) Close() error {
+	return nil
+}
+
+// Read from the RX buffer.
+func (uart UART) Read(data []byte) (n int, err error) {
+	// check if RX buffer is empty
+	size := uart.Buffered()
+	if size == 0 {
+		return 0, nil
+	}
+
+	// only read number of bytes used from buffer
+	for i := 0; i < size; i++ {
+		v, _ := uart.ReadByte()
+		data[i] = byte(v)
```

Comment:
`size` may be bigger than `len(data)` leading to a slice out of bounds error.
Also, `uart.ReadByte()` returns a byte, so this cast is unnecessary.